### PR TITLE
Ubuntu 22 Support Branch

### PR DESCRIPTION
**Pull request by itsarune**
### Description

- Update dependencies and setup_software to support Ubuntu 22
- Force installations on ubuntu 22 to use python3.8 instead of installed python3.10

### Testing Done

- //software/... compiles in ubuntu 22 VM
- thunderscope_main runs in a ubuntu 22 VM

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by Mr-Anyone** [comment]
Interesting. Out of curiosity, what is the intension of updating the python version to 3.11.

**Comment by itsarune** [comment]
yes the main reason is that Ubuntu 22 doesn't have python 3.8 in the system repositories which was our standard for Ubuntu 20. Ubuntu 22 has python3.10 as standard installed python3 version and Ubuntu 20 only has python versions up to python3.9 in the Ubuntu repositories.  I wanted to avoid Ubuntu 22 and Ubuntu 20 using different python versions because there are performance improvements + some features we'd like to use in python 3.9+ (like [type annotations](https://docs.python.org/3/library/typing.html) which become part of the stdlib in python3.9+).

I just decided that the easiest solution would just be installing a python into the /opt/tbotsphython env and everyone would just have the same python version if I compiled from source.

File: environment_setup/setup_software.sh [hunk 1]
```diff
@@ -80,6 +76,7 @@ if [[ $(lsb_release -rs) == "20.04" ]]; then
     host_software_packages+=(llvm-6.0)
     host_software_packages+=(libclang-6.0-dev)
     host_software_packages+=(libncurses5)
+    host_software_packages+=(qt5-default)
```
Comment (nimazareian) [inline environment_setup/setup_software.sh:79]: I thought we were aiming to remove qt5 dependency?

File: environment_setup/setup_software.sh [hunk 2]
```diff
@@ -52,16 +50,14 @@ host_software_packages=(
     protobuf-compiler # This is required for the "NanoPb" library, which does not
                       # properly manage this as a bazel dependency, so we have
                       # to manually install it ourselves
-    python3       # Python 3
-    python3-dev # Python 3 headers
-    python3-venv # Virtual Environment
+    python3.8       # Python 3
+    python3.8-dev # Python 3 headers
+    python3.8-venv # Virtual Environment
```
Comment (nimazareian) [inline environment_setup/setup_software.sh:55]: Were higher versions of Python not simple to integrate? (I noticed the previous comments mentioned 3.11)

File: environment_setup/setup_software.sh (hunk 1 above)
Comment (itsarune) [reply environment_setup/setup_software.sh:79]: I tried to but ErForce uses some deprecated functions that Qt6 doesn't have and I couldn't bother fixing them.

I think the solution is to re-write er-force sim to not use Qt at all and remove the qt dependencies

File: environment_setup/setup_software.sh (hunk 2 above)
Comment (itsarune) [reply environment_setup/setup_software.sh:55]: I tried python3.11 but updating dependencies was annoying so I stuck with 3.8

**Comment by Mr-Anyone** [comment]
Is the performance of thunderscope about the same as before?

**Comment by itsarune** [comment]
I've tried it in a VM and there wasn't any performance downgrade (in fact I think it performed better than my Ubuntu 20 VM).

**Comment by Mr-Anyone** [comment]
> I've tried it in a VM and there wasn't any performance downgrade (in fact I think it performed better than my Ubuntu 20 VM).

I will test this out.

**Review by Mr-Anyone — DISMISSED** [review]
I can confirm that this works. I reinstalled ubuntu 22.04 again and there weren't any performance issues.

File: src/WORKSPACE [hunk 3]
```diff
@@ -99,9 +99,9 @@ protobuf_deps()
 
 git_repository(
     name = "com_github_nelhage_rules_boost",
-    commit = "9f9fb8b2f0213989247c9d5c0e814a8451d18d7f",
+    commit = "4ab574f9a84b42b1809978114a4664184716f4bf",
     remote = "https://github.com/nelhage/rules_boost",
-    shallow_since = "1570056263 -0700",
+    shallow_since = "1694566326 +0000",
```
Comment (itsarune) [inline src/WORKSPACE]: @AdenC123 Can you updated your rules_boost to this commit. This should have the configs you need for linux_aarch64.

**Comment by Mr-Anyone** [comment]
Why downgrade boost? Is it because that the Regrex boost does not compile with g++9?

File: src/software/thunderscope/thunderscope_config.py [hunk 4]
```diff
@@ -51,7 +50,11 @@ def initialize_application():
     app = pyqtgraph.mkQApp("Thunderscope")
 
     # Setup stylesheet
-    apply_stylesheet(app, theme="dark_blue.xml")
+    with open("software/thunderscope/resources/dark_theme.qss", "r") as file:
+        app.setStyleSheet(file.read())
```
Comment (nimazareian) [inline src/software/thunderscope/thunderscope_config.py]: Is this change necessary? Seems like a lot of files in the repo that don't need to be tracked.

Comment (itsarune) [reply src/software/thunderscope/thunderscope_config.py]: I've always had qt theme issues on my laptop and this fixes it by providing all the qt assets exported. Maybe this change makes sense as a separate PR?

**Comment by itsarune** [comment]
@Mr-Anyone The commit is a bit of a misnomer. I've been trying newer versions of boost to help support m1's aarch64 architecture and am trying to find a good version of boost that doesn't require too many software changes from us.

For reference, our current boost version is 1.68 and we're looking to update that to a slightly newer version.

**Review by mkhlb — DISMISSED** [review]
Thunderscope compiles and runs performantly on my end.

File: src/software/thunderscope/thunderscope_config.py (hunk 4 above)
Comment (nimazareian) [reply src/software/thunderscope/thunderscope_config.py]: ya maybe a separate PR. I think it would also make more sense to not have this in our repo, instead, download the files to somewhere in `tbotspython` directory.

**Comment by itsarune** [comment]
Done all my changes for this branch @mkhlb @nimazareian. Review when you can!

**Review by mkhlb — DISMISSED** [review]
Thunderscope builds and runs fine on my end.

**Review by Mr-Anyone — DISMISSED** [review]
seems good.

**Review by nimazareian — APPROVED** [review]
Just upgraded my laptop to Ubuntu 22 and after fixing a broken symlink to gcc (running `sudo ln -sf /usr/bin/gcc-9 /usr/bin/gcc`. The following should also work `sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 9`) everything seem to run as expected.

**Comment by Mr-Anyone** [comment]
also seems to work.

**Review by Mr-Anyone — APPROVED** [review]

**Review by mkhlb — APPROVED** [review]
Tests pass, builds work on my end with Ubuntu 22.
